modules/core/src/lib/layer.ts

```diff
@@ -164,6 +165,18 @@ export default abstract class Layer<PropsT = any> extends Component<PropsT> {
   static defaultProps: any = defaultProps;
   static layerName: string = 'Layer';
 
+  lifecycle: Lifecycle = LIFECYCLE.NO_STATE; // Helps track and debug the life cycle of the layers
+  context: LayerContext | null = null; // Will reference layer manager's context, contains state shared by layers
+  state: Record<string, any> | null = null; // Will be set to the shared layer state object during layer matching
+  internalState: LayerState<PropsT> | null = null;
+
+  constructor(...propObjects: Partial<PropsT>[]) {
+    super(propObjects);
+
+    // Seal the layer
+    Object.seal(this);
```

Comment:
I suppose sealing has its downsides when subclassing and is a lot less valuable in typescript. 

I used to add `seal` mainly to prevent maintainers from slapping on new attributes to layers during run time, which led to unpredictable logic which bit me very hard during a couple of debugging sessions.

But now the typescript compiler will require fields to be declared in the class, so I would be fine with dropping `seal`